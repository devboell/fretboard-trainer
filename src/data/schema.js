import { makeExecutableSchema } from 'graphql-tools'
import resolvers from './resolvers'

const typeDefs = `

  type Quiz {
    name: String
  }

  type Query {
    quizzes: [Quiz]
  }

  schema {
    query: Query
  }
`

export default makeExecutableSchema({ typeDefs, resolvers })
