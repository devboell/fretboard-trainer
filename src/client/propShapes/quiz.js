import pt from 'prop-types'


export const panelModeShape = pt.shape({
  id: pt.string,
  question: pt.string,
  answer: pt.string,
})

export const quizShape = pt.shape({
  id: pt.string,
  name: pt.string,
  type: pt.string,
  tuning: pt.string,
  width: pt.number,
  panelModes: pt.arrayOf(panelModeShape),
})
