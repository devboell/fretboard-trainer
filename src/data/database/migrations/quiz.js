export function up(knex, Promise) {
  return Promise.all([
    knex.schema.createTable('Quiz', (table) => {
      table.increments('id').notNullable()
      table.string('name').notNullable()
      table.enum('type', ['pc']).notNullable()
    }),
  ])
}

export function down(knex, Promise) {
  return Promise.all([
    knex.schema.dropTable('Quiz'),
  ])
}
