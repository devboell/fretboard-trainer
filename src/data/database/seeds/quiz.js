const quizzes = `
  INSERT INTO 'Quiz' VALUES
  (
    1,
    'Notes I',
    'Very easy training to help memorize the notes on the fretboard. The fretboard displays all the note names, so they are easy to find. This is just a warm-up.',
    'pc',
    'standard',
    13,
    0,
    1,
    0,
    0,
    1,
    0
  ),
  (
    2,
    'Notes II',
    'no description',
    'pc',
    'standard',
    13,
    1,
    1,
    1,
    5,
    0,
    0
  ),
  (
    3,
    'quiz three',
    'no description',
    'pitch',
    'standard',
    13,
    1,
    1,
    1,
    5,
    0,
    1
  ),
  (
    4,
    'quiz four',
    'no description',
    'pitch',
    'standard',
    13,
    1,
    1,
    1,
    5,
    0,
    0
  ),
  (
    5,
    'quiz five',
    'no description',
    'pitch',
    'standard',
    13,
    1,
    1,
    1,
    5,
    0,
    0
  ),
  (
    6,
    'quiz six',
    'no description',
    'pitch',
    'standard',
    13,
    1,
    1,
    1,
    5,
    0,
    0
  )
`

export function seed(knex) { // eslint-disable-line import/prefer-default-export
  return knex('Quiz').truncate()
    .then(() =>
      knex.raw(quizzes))
}
