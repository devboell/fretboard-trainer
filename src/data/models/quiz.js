import knex from '../connector'

export const findAll = () =>
  knex('Quiz').select()

export const findById = id =>
  knex('Quiz').where({ id }).first()

export const panelModesByQuizId = id =>
  knex('Quiz')
    .select(['pm.id', 'pm.question', 'pm.answer'])
    .leftJoin('Quiz_PanelMode as qpm', 'Quiz.id', 'qpm.quizId')
    .leftJoin('PanelMode as pm', 'qpm.panelModeId', 'pm.id')
    .where({ 'Quiz.id': id })

export const create = async (args) => {
  const result = await knex('Quiz').insert(args)
  return result[0]
}

export const update = (id, values) =>
  knex('Quiz')
    .where({ id })
    .update(values)


export const remove = id =>
  knex('Quiz')
    .where({ id })
    .del()
