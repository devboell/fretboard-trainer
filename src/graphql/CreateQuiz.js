import gql from 'graphql-tag'
import QUIZ_FRAGMENT from './QuizFragment'

export default gql`
    mutation CreateQuiz(
    $input: CreateQuizInput,
  ) {
    createQuiz(
      input: $input
    ) {
      ...QuizFragment
    }
  }
  ${QUIZ_FRAGMENT}
`
