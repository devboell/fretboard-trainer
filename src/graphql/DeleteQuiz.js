import gql from 'graphql-tag'

export default gql`
  mutation DeleteQuiz($id: ID) {
    deleteQuiz(id: $id)
  }
`
