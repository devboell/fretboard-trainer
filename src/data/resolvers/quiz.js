import * as Quiz from '../models/quiz'
import * as QPM from '../models/quizPanelMode'


export const quizzes = async () =>
  Quiz.findAll()

export const panelModes = async obj =>
  Quiz.panelModesByQuizId(obj.id)

export const createQuiz = async (_, args) => {
  const { quiz, panelModeIds } = args.input
  const newId = await Quiz.create(quiz)
  await QPM.create(newId, panelModeIds)
  return Quiz.findById(newId)
}

export const updateQuiz = async (_, args) => {
  const {
    id,
    quiz,
    panelModeIds,
  } = args.input

  await Quiz.update(id, quiz)
  await QPM.removeByQuizId(id)
  await QPM.create(id, panelModeIds)

  return Quiz.findById(id)
}

export const deleteQuiz = async (_, args) => {
  const nrOfRows = await Quiz.remove(args.id)

  return nrOfRows === 1 ? args.id : 'ERROR' // TODO: unhandled error
}
