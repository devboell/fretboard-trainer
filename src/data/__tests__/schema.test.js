import { graphql } from 'graphql'
import knex from 'data/connector'
import QUIZZES from 'graphql/Quizzes'
import PANEL_MODES from 'graphql/PanelModes'
import CREATE_QUIZ from 'graphql/CreateQuiz'
import UPDATE_QUIZ from 'graphql/UpdateQuiz'
import DELETE_QUIZ from 'graphql/DeleteQuiz'
import * as fxt from 'fixtures/graphql/quiz'
import schema from '../schema'


describe('schema', () => {
  setUpData(knex)

  it('should return quizzes', async () => {
    const queryObj = {
      schema,
      source: QUIZZES.loc.source,
    }
    const result = await graphql(queryObj)
    expect(result.data.quizzes).toEqual(fxt.quizzes)
  })

  it('should return panelModes', async () => {
    const queryObj = {
      schema,
      source: PANEL_MODES.loc.source,
    }
    const result = await graphql(queryObj)
    expect(result.data.panelModes).toMatchSnapshot()
  })

  it('should create a quiz', async () => {
    const queryObj = {
      schema,
      source: CREATE_QUIZ.loc.source,
      variableValues: fxt.createQuizInputValues,
    }
    const result = await graphql(queryObj)
    expect(result.data.createQuiz).toEqual(fxt.createdQuiz)
  })

  it('should update a quiz', async () => {
    const queryObj = {
      schema,
      source: UPDATE_QUIZ.loc.source,
      variableValues: fxt.updateQuizInputValues,
    }
    const result = await graphql(queryObj)
    expect(result.data.updateQuiz).toEqual(fxt.updatedQuiz)
  })

  it('should delete a quiz', async () => {
    const queryObj = {
      schema,
      source: DELETE_QUIZ.loc.source,
      variableValues: { id: 3 },
    }
    const result = await graphql(queryObj)
    expect(result.data.deleteQuiz).toBe('3')
  })
})
