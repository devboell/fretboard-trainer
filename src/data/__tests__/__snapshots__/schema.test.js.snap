// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema should return panelModes 1`] = `null`;
